### ISSUE TITLE
Exploration: How to handle deletes when all workplacement write to the same directory

### ISSUE THREAD

**Issue by aclevername** [syntasso/kratix#132]
WIP coming soon

As part of https://github.com/syntasso/kratix/issues/127 we are adding support for having Destination receives all files from pipelines in the same place. The problem is how do we handle deleting files from this. Today we just `rm -rf <sub-dir-for-a-given-request>/` because we knew everything was always nested under a directory. Now we have a shared place for the files to live this becomes more complex.

## Setup
Resource Request A outputs 
```
/kratix/output/foo.yaml
/kratix/output/bar.yaml
```

and resource request B outputs
```
/kratix/output/banana.yaml
/kratix/output/apple.yaml
```

And they are scheduled to a destination that has `.spec.filepath.mode=none` then the state store should end up looking like:

```
foo.yaml
bar.yaml
banana.yaml
apple.yaml
``` 

## Updates

Resource Request A is updated and now outputs:
```
/kratix/output/foo.yaml
```

No longer `bar.yaml`. The problem comes when we reconcile this new change, *how do we know that we should delete `bar.yaml`?* Kubernetes is [level based](https://stackoverflow.com/questions/31041766/what-does-edge-based-and-level-based-mean), meaning that at any point you should be able to look at *the most up to date kubernetes resource* and figure out what the desired state is. However in this scenario in order to delete `bar.yaml` we have to know about its historical existence (we are now [edge based](https://stackoverflow.com/questions/31041766/what-does-edge-based-and-level-based-mean)), which we don't know.

## Coming soon
some ideas on how to solve it, probably in a miroboard.

**Comment by aclevername** [comment]
After some exploration we have decided to down the following approach outlined in the diagram below.

TLDR:
- We will create a `.kratix/` directory in the statestore when `destination.spec.filepath.mode=none`. `filepath.mode` defaults to `nestedByMetdata` which is the current default behaviour in kratix to date.
- This will contain a file for each workplacement, inside that file it will list what files in the statestore belong to that workplacement
- We will use this file to give us information when deleting files during updates/deletes of workplacement

![Screenshot 2024-05-22 at 11 40 35](https://github.com/syntasso/kratix/assets/19669095/b8b00ca1-ed07-4d24-8641-ee7900f3b6c8)
